### ISSUE TITLE
Non-disruptive Upgrades

### ISSUE THREAD

**Issue by MTRichards** [owncloud/core#18721]
Goal:
- Upgrade NEVER leads to a non working ownCloud installation. (ownCloud should never be in a state where it is not working / can't be recovered)
- Done very fast (1min) (long running steps should happen before or after asynchronously)
- Works from CLI and web
- Always tells the admin whats happening now, what happens next and how long it will take.
- Offers roll-back

Approach:
- New isolated updater, all code files are located in /updater, no interaction with the rest of ownCloud. 
- No code, configuration, database or anything else used from core is required. (- included with normal ‘required’ to make it isolated)
- Shipped with ownCloud core
- Can be called by CLI or web (updater/update.php or  /updater)
- Offer interactive or automatic mode (interactive recommended)
- Offers to do a /data backup if possible
- Offers to do a db backup if possible
- Offers to do a full rollback if failed or offers a Retry if a steps fails temporarily.
- Support the admin with cleaning caches/backup
- Can detect failed app upgrades and can move failed apps into a ‘incompatible-apps’ folder
- Included tipps and tricks and hints 
- Included links to documentation for more information
- Upgrade steps should be split up into pre/post steps and optional/required steps. For example mimetype conversion can happen when an user requests his files instead of doing 16 hour queries on update.
- ownCloud should check itself for integrity problems. So it ownCloud and the DB schema are unmodified or changed.
- It should be possible to export schema migration statements to do the DB migration asynchronously.
  Currently: On-the fly schema migration, diff
  Wanted: Have predictable database upgrade changes, move to define the migration steps 
- DOES NEVER BREAK

Concrete process:
- tbd

Todos:
- Check how wordpress handles it
- Check how Piwik handles it
- Go through existing issues to see if we catch all problems
- somehow make sure that the admin doesn't miss new config options

**Comment by RealRancor** [comment]
Ref to an older issue at the documentation tracker: https://github.com/owncloud/documentation/issues/394

**Comment by Xenopathic** [comment]
Two parts to an upgrade: code upgrade and DB upgrade.

Code upgrades need to ensure the server only sees one of two states - full old code or full new code. If some classes are new while others are still not updated then things will break. The ways to do this are with filesystem snapshots and atomic upgrades (supported by btrfs and other modern filesystems), where a snapshot of the filesystem is taken, changes made to it while the PHP process still sees the snapshot, then updating the snapshot to reflect the changed files.

DB upgrades are somewhat more difficult. I imagine the easiest thing to start with is setting oC to read-only, so no DB writes can occur. This would greatly help keeping things consistent. Then, just like code upgrades, the schema changes need to be atomic. I think transactions can help here? In fact, ideally schema updates and repair steps would all happen in one atomic operation, so the DB goes from pre-upgrade to post-upgrade and no running instance sees anything in between. I'm not sure how feasible this is though.

Finally is the problem of synchronising code upgrades and DB upgrades. When the code is upgraded but the DB is still old, ownCloud cannot work. Likewise the other way around. Perhaps the best way to handle this is to prepare filesystem snapshots on each server, perform the atomic database upgrade, then apply the filesystem snapshots. The only downtime during this would be after the database upgrade but before the filesystem snapshot update, which should be short as applying snapshots is usually very fast. Unfortunately the only way I see online DB upgrades working properly is by only allowing read-only access for the short duration.

**Comment by DeepDiver1975** [comment]
Code update is of zero problem: web Server nodes to be updated will be taken offline.

Anything else is highly critical. No idea if doable a all. 

I honestly question where this requirement is coming from. Even in the investment banking sector e.g. a downtime of the financial systems was no Problem. 

What we should AIM for is a more managable upgrade process which is actually testable and reproducable.

Let me once more reference doctrine migrations to be used instead of this mixture of Schema diffing magic plus wild php scripting.

**Comment by zijlstra-it** [comment]
I'd like to add to this that being able to "produce" a complete DB change script, so not only the schema changes, would be very helpful to being able to test and time the databases updates on a test instance of the database. 
For me it's unfeasible to have a test environment with the same number of accounts and data, but a database copy is no problem and a system with the same specs either, so if I would be able to run ALL database changes for an upgrade on a test DB, I can time the process and test if it works at all in our case.
I used the "db:generate-change-script" to test our upgrade to 7.06 from 6.06 but it only covered the schema changes, not the table updates due to changed behaviour of OC, e.g. the "Shared" folder which changed from 6 to 7.

**Comment by schnidrig** [comment]
Preparing a minor update from 7.0.8 to 7.0.10 I took a closer look at what happens with the database. Here is what I found:

- the schema did not change: db_structure.xml is more than a year old.
- occ db:generate-change-script produces a lot of output. 
  Here a small but very disturbing extract from the output:

      CREATE TEMPORARY TABLE __temp__oc_filecache AS SELECT fileid, storage, path, path_hash, parent, name, mimetype, mimepart, size, mtime, storage_mtime, encrypted, unencrypted_size, etag, permissions FROM oc_filecache;
      DROP TABLE oc_filecache;
      CREATE TABLE oc_filecache (fileid INTEGER NOT NULL, storage INTEGER DEFAULT 0 NOT NULL, path VARCHAR(4000) DEFAULT NULL, path_hash VARCHAR(32) DEFAULT '' NOT NULL, parent INTEGER DEFAULT 0 NOT NULL, name VARCHAR(250) DEFAULT NULL, mimetype INTEGER DEFAULT 0 NOT NULL, mimepart INTEGER DEFAULT 0 NOT NULL, mtime INTEGER DEFAULT 0 NOT NULL, storage_mtime INTEGER DEFAULT 0 NOT NULL, encrypted INTEGER DEFAULT 0 NOT NULL, etag VARCHAR(40) DEFAULT NULL, permissions INTEGER DEFAULT 0, "size" INTEGER DEFAULT '0' NOT NULL, "unencrypted_size" INTEGER DEFAULT '0' NOT NULL, PRIMARY KEY(fileid));
      INSERT INTO oc_filecache (fileid, storage, path, path_hash, parent, name, mimetype, mimepart, "size", mtime, storage_mtime, encrypted, "unencrypted_size", etag, permissions) SELECT fileid, storage, path, path_hash, parent, name, mimetype, mimepart, size, mtime, storage_mtime, encrypted, unencrypted_size, etag, permissions FROM __temp__oc_filecache;
      DROP TABLE __temp__oc_filecache;
      CREATE INDEX fs_storage_size ON oc_filecache (storage, size, fileid);
      CREATE INDEX fs_storage_mimepart ON oc_filecache (storage, mimepart);
      CREATE INDEX fs_storage_mimetype ON oc_filecache (storage, mimetype);
      CREATE INDEX fs_parent_name_hash ON oc_filecache (parent, name);
      CREATE UNIQUE INDEX fs_storage_path_hash ON oc_filecache (storage, path_hash);

Why on earth is this necessary? Our oc_filecache table holds more than 20 Mio rows. Please, if the schema doesn't change, then just leave these large tables alone!

**Comment by MTRichards** [comment]
https://github.com/owncloud/core/pull/18599#issuecomment-141484384

**Comment by karlitschek** [comment]
proposal for the concrete upgrade process:
# upgrade++ NG 2.0

- isolated upgrade in /upgrade
- Web and CLI
- All steps can be executed seperately with CLI options
- default is interactive mode 
- write everything into an upgrade.log
- write a status file to make resume possible.

Other:
- Make it possible to export a DB migration sctipt



## upgrade steps

- Info. Show what is going to happen with explanation. 
- System check. System health and if dependencies are OK (we also count the number of files and DB entries and give time estimations based on hardcoded estimation)
- maintainance mode on

Detect
	- 1. currently existing code, 
	- 2. version in config.php, 
	- 3. online available verison.
(ASK) what to do? (download, upgrade, abort, …)

in case of download: (
	- download new ownCloud release
	- backup of old code (optionally)
	- unpack tar and replace code
	- exit and tell user to please restart upgrade script
) 


- Backup DB (optionally)

- Backup data (optionally)

- Repair and cleanup (pre upgrade, DB collations update, ..) [danger, might take long]

- db schema upgrade simulated (optionally, can be done online in advance)

- db schema upgrade real
[danger, might take long]


- disable 3rdparty/no shipped apps 


- execute core upgrade scripts
[danger, might take long]

- upgrade shipped apps [danger, might take long]
	- schema migration
	- app upgrade scripts


- try reenable and upgrade 3rdparty/non shipped apps. (one app after another)


- clean caches

- repair and cleanup step 1 (post upgrade, repair legacy storage, ..) [danger, might take long]

- restart webserver (ask, propably a good idea, just to be sure)

- update config.php 

- maintainance mode off

- repair and cleanup step 2 (online) [danger, might take long]




## Meta todos for upcoming releases:
- DB: no more table/column changes. Only add new tables/columns
- Repair scripts: make possible to run in production
- migration scripts: handle half migrated status in production

**Comment by MorrisJobke** [comment]
Copy of the notes from our session at the conf - @karlitschek feel free to merge this with you comment:

### Goals

 * Never kill OC
 * ownCloud should never be in a state where it is not working / can't be recovered
 * Fast
 * Interactive
   * No user feedback, should have more verbosity => How long does each step take / estimates / …
   * Offer advises such as "Please create a database backup"
   * Morris: Interactivity will create problems for enterprise deployments, more verbosity would be good. => Where is the problem? Deployment problem or ownCloud bug?
   * For example when occ runs show what will run and offer possibility to skip some migrations
 * Wordpress / Piwik-Reliability

# Ideas

 * Isolated updater code
   * Self-contained, so that it does not shoot itself into the foot by accident. Such as replacing files that it depends on and then clearing OPCache.
 * CLI / WEB
  * share as most as possible code paths
 * App Enable Check
   * a.k.a disable apps that break stuff
   * Lukas: Not moving apps to "/incompatible-apps" due to security concerns ("Enterprises do want have a Read-only oC")
 * Pre/Post steps
   * For example mimetype conversion can happen when an user requests his files instead of doing 16 hour queries on update.
   * Have:
     * Pre-steps, before the update, required
     * Post-steps, after the update, can be done while it is running in production
     * Optional (maybe)
 * Caches
   * Ensure that all caches are cleaned (memcached / APCu / …)
   * => Already done by prepending the caching keys with the version
 * Backups
 * Handholding
   * Help users more, such as explaining how to restart their web server
 * Time estimations (see "Interactive" from "Goals")
 * Progress (see "Interactive" from "Goals")
   * Update steps might take from hours and there is no feedback, some customers believe these were bugs and stopped the update => Broken instances
 * Upgrading apps
   * Exception in  apps that aren't tested properly/"interesting configuration" -> kills the whole upgrade run
 * Rollback on failure
   * Half-way migrated, something is missing, no rollback for this
   * Unlikely to be implemented, very hard to do right, can be done with snapshots / backups.
 * Integrity problems
   * some files are not upgraded
     * Verify if the code that is currently running, is exactly the code that is running (possible candidate: .htaccess because of hidden file)
   * Database
     * People are applying random patches to database scheme
 * App disabled during upgrade
   * Re-enable => Kills instance because of upgrade page
     * Enable an app => Get upgrade mode, enable another one via Ajax => Fails since in upgrade mode
 * Proper schema migration
   * Currently: On-the fly schema migration, diff
   * Wanted: Have predictable database upgrade changes, move to define the migration steps
 * Partial upgrade for some users/migration of users from old version instance to a new version instance owncloud/enterprise#686
   * Wanted: Have a test group that uses new app
     * Not possible to update that easily
     * Jörns idea: Have an old and new version and move users, for sharing use federated sharing

**Comment by PVince81** [comment]
Note regarding pre/post steps: everything that is done "lazily" needs to be properly protected by locking to make sure the migration isn't done twice and causing unpredictable results.
Same for anything that is done "at login time", considering that multiple logins can happen at the same time.

**Comment by PVince81** [comment]
Also to note:
- the current app architecture doesn't allow for generating SQL statements because apps can provide generic PHP code in "appinfo/update.php" which is simply run when updating the app. To fix this, the apps would need to be told that we only want to get the SQL from the possible update and not actually perform it. Or deprecate "appinfo/update.php" and expect apps to implement the RepairStep interface for better compliance with the process. (currently only core can provide/register RepairStep instances)

- some repair steps are not running SQL queries directly but using OC core code which contains the logic necessary. We might need to decouple the repair steps from the core code, if possible. (and make this a guideline when writing new repair steps)

- some repair steps might operate directly on files (not purely SQL/DB changes). I'm not aware of any that does so at the moment: the encryption app usually requires the admin to run an additional "./occ" command to operate on files.

**Comment by PVince81** [comment]
As discussed previously with @schiesbn, long running repair steps (or on-demand) might work, but they would also need to be able to run in many future versions of OC.
Consider the following scenario
- setup is currently on OC 9.0
- admin decides to update to OC 12.0 (without skipping major versions)
- repair step "FixSomething" is run in the background for all users without disrupting functionality
- whenever a user accesses a folder, "FixMimeType" is applied (for example)
- admin upgrades to OC 9.1
- repair step "FixSomething" hasn't finished, so need to continue running
- new repair step "FixSomethingAfterFixSomething" (needs "FixSomething") to finish first
- "FixMimeType" hasn't been run for all files because not every user has viewed all their (unchanged) folders
- admin upgrades to OC 9.2
- repair step "FixSomething" still running
- ...
- admin upgrades to OC 12
- at this point, the repair steps should still be able to continue running, including the "lazy" fixes like "FixMimeType".

This is only a potential issue that can happen, so we need to consider that when reaching OC 12, some database entries might still be in the state pre-OC 9, so the code that repairs them needs to remain in place. (and potentially need to be retested, hopefully automatically)

Basically, this brings the more detailed requirement:
- old repair steps (code) still need to be available in the newest versions and be able to run / continue running
- lazy/on-demand repair steps need to work in the newest versions
- always assume that the DB might have unrepaired entries from older updates

**Comment by VicDeo** [comment]
what make the life painful right now:
- users may have some unrelated items in the webserver root directory (webalazer logs, etc) or other places. Which makes it possible to detect whether file belongs to the owncloud or not.
Some sort of a centralized storage that knows exact list of files for any version of owncloud
- movable 3rdparty dir and several apps directories (writable/not writable)

**Comment by PVince81** [comment]
- making the web based updater is tricky if we want it to spawn processes when driving the update, popen() and exec() might be blocked in shared hosters which are the main target for such a web UI

**Comment by VicDeo** [comment]
Wordpress reliability... hm
I would say https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/upgrade.php#L478

**Comment by MorrisJobke** [comment]
> Wordpress reliability... hm
> I would say https://github.com/WordPress/WordPress/blob/master/wp-admin/includes/upgrade.php#L478

:+1:

**Comment by PVince81** [comment]
Wordpress also has update issues, just read the tweets https://twitter.com/search?q=wordpress%20update%20break&src=typd

Anyway, comparing to Wordpress is not the point. The point is simply to make OC updates more reliable.

**Comment by DeepDiver1975** [comment]
> Anyway, comparing to Wordpress is not the point. The point is simply to make OC updates more reliable.

indeed - but there was the statement: "Other do upgrade without problems - like wordpress ..."

**Comment by MorrisJobke** [comment]
> UPD I wonder how much time these steps will take for e.g. 1000 users having 10000 posts...

It's simply not possible to compare Wordpress&Piwik with ownCloud because the amount of data are completely different. We need to find our way to make it reliable.

**Comment by DeepDiver1975** [comment]
@karlitschek @cmonteroluque @PVince81 regarding the location of the new tool which is under development: I propose to put it into the existing updater app repo.

The updater app as we ship it until 8.2 will no longer be used and the code can continue to live in the stableX branches. Master will hold the new tool.

Objections?

**Comment by karlitschek** [comment]
Works for me. We will move it into /updater of the owncloud root in the tar files release.

**Comment by DeepDiver1975** [comment]
> Works for me. We will move it into /updater of the owncloud root in the tar files release.

THX - @VicDeo please prepare a PR these days - THX

**Comment by PVince81** [comment]
It would be good to raise a separate ticket for the requirements/acceptance criteria that are scoped for 9.0 to get a clearer picture. I thought we had a list already, can't seem to find it.
Thanks.

**Comment by DeepDiver1975** [comment]
@PVince81 just to document the current apporach:
- the first implementation will be console based only
- capabilities for web based user interaction is being thought off during implementation of the first step
- steps to be implemented are https://github.com/owncloud/core/issues/18721#issuecomment-143215159
- the tool will be developed isolated from owncloud core as self contained web and console project
- it is expected to have the tool running in a webspace root in a subfolder 'updater'
- all steps in it's first implementation will hold simple test out put to tell the user what to do :wink: 
- interaction with the existing/installed/updated core is to be done via the occ command line tool (unclear in case of web based approach)

@VicDeo did I miss anything? THX

**Comment by cmonteroluque** [comment]
@DeepDiver1975 thanks

**Comment by VicDeo** [comment]
@DeepDiver1975 Looks like the description is complete

**Comment by VicDeo** [comment]
Notes: right now I simply add `.md5` postfix to the download URL to obtain the package checksum location. 
But I really appreciate https://github.com/owncloud/core/issues/20056 ;)

**Comment by DeepDiver1975** [comment]
> it's the right time to start preparing the filelists for each and every released versions

@LukasReschke this sounds like a reuse case of the signing mechanism

**Comment by DeepDiver1975** [comment]
> target PHP version?

5.4, 5.5, 5.6, 7

**Comment by DeepDiver1975** [comment]
> It would be great to have a certain non zero exit code for the case when occ was run by improper user https://github.com/owncloud/core/blob/master/console.php#L63
> Otherwise I would stop the process with something like
> 'Can not parse occ response, it says: raw_output_here' because it is not possible to determine what exactly is wrong.

pull request welcome :wink:

**Comment by VicDeo** [comment]
@DeepDiver1975 
> @LukasReschke this sounds like a reuse case of the signing mechanism

 I don't want to nag user for using something unsigned.
What I need is a complete list of files that I am allowed to remove.
And I should admit  Wordpress has a  perfect approach here: there are 3 directories and a dozen of files.
https://github.com/WordPress/WordPress

No multiple app directories, no movable 3rdparty or other sexy features leading to unpredictable results.

**Comment by DeepDiver1975** [comment]
The full list of files which are contained in a release are part of the signing mechanism.
This should give a good starting point.

**Comment by ferdiga** [comment]
please don't allow to click "upgrade" in the UI if the amount of data (oc_filecache) is to big, as it  will run a probably useless migration-test and run into a timeout .....
it just happened to me after activating "Gallery" in 8.2.1 which apparently requires an "upgrade".

**Comment by PVince81** [comment]
@VicDeo what's the status of this ? I see there were unfinished checkboxes in the merged PR https://github.com/owncloud/updater/pull/187#issue-117360434.
Would be good to open a "task" ticket with a list of checkboxes of what's left to do.

Thanks.

**Comment by PVince81** [comment]
Just thought about something: what about servers that have no access to the Internet, how could they benefit from the updater ? I think there should be an option to specify a custom URL for downloading the tarball. An admin can then either put "file:///" or the URL of a local intranet resource that will deliver the new tarball to the instance.

Even simpler would be to be able to pass the path of the tarball when running the upgrade command.

This would also make it possible to test upgrades more easily by building "dummy" tarballs for the next version instead of requiring a remote server to deliver test tarballs.

**Comment by PVince81** [comment]
Basic version of the updater CLI + web UI was implemented.

@VicDeo can you post a link to the summary of what was done so far ?

We can then move this ticket to 9.1 and/or raise separate tickets for the remaining tasks.

**Comment by cmonteroluque** [comment]
aligned with 
https://github.com/owncloud/updater/issues/250

**Comment by MTRichards** [comment]
https://github.com/owncloud/appstore-issues/issues/4

**Comment by VicDeo** [comment]
@MTRichards https://github.com/owncloud/appstore-issues/issues/4 is addressed in https://github.com/owncloud/administration-internal/pull/12 but I'm not sure what decision was made about it.

**Comment by MTRichards** [comment]
Estimation: 6 weeks

cc:/ @PVince81

**Comment by PVince81** [comment]
moving topic to 9.2.

The separated updater has already improved a lot.

**Comment by PVince81** [comment]
I think the only stuff in scope for 10.0 right now is making the updater app be able to upgrade over multiple versions. And that will only work starting with 10.0 once Doctrine migrations are used for any upgrades.

@DeepDiver1975 @VicDeo correct ?

**Comment by PVince81** [comment]
I'm moving this to backlog.

Will need to be rescheduled once we intend to work on the other remaining tasks. (there are lots of them, will likely need separate estimations and scheduling)

@DeepDiver1975 @butonic @hodyroff @pmaier1

**Comment by PVince81** [comment]
removing p1 as this is not in focus currently
